hazelcast/src/test/java/com/hazelcast/test/bounce/BounceMemberRule.java

```diff
@@ -0,0 +1,514 @@
+
+        testRunning.set(true);
+    }
+
+    private void tearDown() {
+        try {
+            if (taskExecutor != null) {
+                taskExecutor.shutdown();
+            }
+            // shutdown test drivers first
+            if (testDrivers != null) {
+                for (HazelcastInstance hz : testDrivers) {
+                    hz.shutdown();
+                }
+            }
+            if (factory != null) {
+                factory.shutdownAll();
+            }
+        } catch (Throwable t) {
+            // any exceptions thrown in tearDown are not interesting and may hide the actual failure, so are ignored
+        }
+    }
+
+    // reflectively instantiate client-side TestHazelcastFactory
+    private TestHazelcastInstanceFactory newTestHazelcastFactory() {
+        try {
+            Class klass = ClassLoaderUtil.loadClass(null, "com.hazelcast.client.test.TestHazelcastFactory");
+            return (TestHazelcastInstanceFactory) klass.newInstance();
+        } catch (Exception e) {
+            throw new AssertionError("Could not instantiate client TestHazelcastFactory: " + e.getMessage());
+        }
+    }
+
+    /**
+     * @return a Statement which starts a separate thread that bounces members of the cluster until {@link #testRunning} is
+     * {@code false}, then proceeds with the given {@code next} Statement.
+     */
+    public final Statement startBouncing(final Statement next) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                bouncingMembersTask = spawn(new MemberUpDownMonkey(members));
+                next.evaluate();
+            }
+        };
+    }
+
+    /**
+     * @param next Statement to evaluate before stopping bouncing members
+     * @return a Statement which executes the {@code next} Statement, then sets the test's running flag to false and waits for
+     * the member bouncing thread to complete.
+     */
+    public final Statement stopBouncing(final Statement next) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                try {
+                    next.evaluate();
+                } finally {
+                    testRunning.set(false);
+                    try {
+                        bouncingMembersTask.get();
+                    } catch (Exception e) {
+                        // ignore
+                    }
+                }
+            }
+        };
+    }
+
+    // wait until test tasks complete or throw exception
+    private void waitForFutures(Future[] futures) {
+        boolean hasTestFailed = testFailed.get();
+        for (Future future : futures) {
+            try {
+                // TODO check logic here
+                // if the test failed, try to locate immediately the future that is done and will throw an exception
+                if ((hasTestFailed && future.isDone()) || !hasTestFailed) {
```

Comment:
I updated the future-waiting code to avoid the edge case where `testFailed.get()` might return `true` while its `Future` might not be done yet.